src/dev/lupio/lupio_rng.hh

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __DEV_LUPIO_LUPIO_RNG_HH__
+#define __DEV_LUPIO_LUPIO_RNG_HH__
+
+#include <random>
+
+#include "debug/LupioRNG.hh"
+#include "dev/io_device.hh"
+#include "params/LupioRNG.hh"
+#include "sim/system.hh"
+
+namespace gem5
+{
+
+using namespace RiscvISA;
+
+/**
+ * LupioRNG:
+ * A Random Number Generator virtual device that returns either a random
+ * value, or a seed that can be configured by the user.
+ */
+class Platform;
+
+class Plic;
```

Comment:
Related to question above: is there a way not to be tied to the PLIC specifically? Can't devices see the interrupt controller they're attached to as a generic component that offers the necessary features?

Reply:
Follow-up:

Does the RNG need to have a handle on the PLIC or could it just have a handle on a `BasicPioDevice`? In the latter case, it would make it compatible with any interrupt controller that derives from `BasicPioDevice` (which seems to be the case for all memory-mapped platforms).

Could you try (in that case, instead of declaring `Plic *plic`, you would just declare `BasicPioDevice *pic`), or investigate if it's possible?